### PR TITLE
update keen home smart vent device handler

### PR DESCRIPTION
**Pull request by mcous**
This PR updates the Device Handler for the Keen Home Smart Vent to a production ready release. Compared to the currently released handler, this version mostly cleans up some code and fixes obstruction handling:
- Does not allow level or on/off to be changed while a vent is obstructed 
- Properly handles UI state for an obstructed vent
- Allows a user to tell a vent to recalibrate and attempt to clear an obstruction wirelessly

### CODE REVIEW

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 1]
```diff
@@ -327,35 +311,79 @@ private def makeSerialResult(serial) {
         value: serial,
         descriptionText: "${linkText} has serial ${serial}" ]
 }
-/**** COMMAND METHODS ****/
-// def mfgCode() {
-// 	 ["zcl mfg-code 0x115B", "delay 200"]
-// }
 
+// takes a level from 0 to 100 and translates it to a ZigBee move to level with on/off command
+private def makeLevelCommand(level) {
```
Comment (tpmanley) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:316]: We've been improving the API and there is now a method `zigbee.setLevel(level)` that does the same thing as this method and should be used instead

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 2]
```diff
@@ -327,35 +311,79 @@ private def makeSerialResult(serial) {
...
+    log.debug "open ${linkText}"
+
+    // only change the state if the vent is not obstructed
+    if (device.currentValue("switch") == "obstructed") {
+        log.error("cannot open because ${linkText} is obstructed")
+        return
+    }
+
     sendEvent(makeOnOffResult(1))
     "st cmd 0x${device.deviceNetworkId} 1 6 1 {}"
```
Comment (tpmanley) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:339]: Use `zigbee.on()` instead

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 3]
```diff
@@ -327,35 +311,79 @@ private def makeSerialResult(serial) {
...
+    log.debug "close ${linkText}"
+
+    // only change the state if the vent is not obstructed
+    if (device.currentValue("switch") == "obstructed") {
+        log.error("cannot close because ${linkText} is obstructed")
+        return
+    }
+
     sendEvent(makeOnOffResult(0))
     "st cmd 0x${device.deviceNetworkId} 1 6 0 {}"
```
Comment (tpmanley) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:353]: Use `zigbee.off()` instead

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 4]
```diff
@@ -363,29 +391,26 @@ def setLevel(value) {
...
 def getOnOff() {
     log.debug "getOnOff()"
 
+    // disallow on/off updates while vent is obstructed
+    if (device.currentValue("switch") == "obstructed") {
+        log.error("cannot update open/close status because ${getLinkText(device)} is obstructed")
+        return []
+    }
+
     ["st rattr 0x${device.deviceNetworkId} 1 0x0006 0"]
```
Comment (tpmanley) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:407]: Use `zigbee.onOffRefresh()` instead

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 5]
```diff
@@ -395,12 +420,13 @@ def getPressure() {
...
-    // 0 = attribute within cluster
-    // sendEvent(name: "level", value: value)
+
+    // disallow level updates while vent is obstructed
+    if (device.currentValue("switch") == "obstructed") {
+        log.error("cannot update level status because ${getLinkText(device)} is obstructed")
+        return []
+    }
+
     ["st rattr 0x${device.deviceNetworkId} 1 0x0008 0x0000"]
```
Comment (tpmanley) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:430]: Use `zigbee. levelRefresh()` instead

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 6]
```diff
@@ -327,35 +311,79 @@ private def makeSerialResult(serial) {
...
+        value: "clearing",
+        descriptionText: "${linkText} is clearing obstruction"
+    ])
+
+    // send a move command to ensure level attribute gets reset for old, buggy firmware
+    // then send a reset to factory defaults
+    // finally re-configure to ensure reports and binding is still properly set after the rtfd
+    [
+        makeLevelCommand(device.currentValue("level")), "delay 500",
+        "st cmd 0x${device.deviceNetworkId} 1 0 0 {}", "delay 5000"
```
Comment (tpmanley) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:371]: Use `zigbee.zigbeeCommand(0, 0, "{}")` instead. The signature is `zigbeeCommand(cluster, attributeId, values)`

**Comment by mcous** [comment]
@tpmanley Is there documentation you can send me for the ZigBee commands you've referenced? The [SmartThings documentation page](http://docs.smartthings.com/en/latest/ref-docs/device-handler-ref.html#zigbee) just reads "The documentation for this property is incomplete."

**Comment by tpmanley** [comment]
The API improvements and documentation updates are still a work in progress, but you can look at https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/zigbee-dimmer.src/zigbee-dimmer.groovy for examples of these method in use.

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy (hunk 1 above)
Comment (mcous) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy:316]: On our device, a (ZigBee) level of 0xFF is a reserved value, which I don't think the `zigbee.setLevel` method would be able to take into account. Also, I don't know what command `setLevel` sends, but for our device it's important the command is a "move to level with on/off" rather than a "move to level".

**Comment by mcous** [comment]
I'm hesitant to update this handler to use undocumented / "work in progress" API commands, especially when we know what we've currently got works and we've encountered hub incompatibilities with certain commands in the past.

Is there an ETA on that documentation or those features?

**Comment by tpmanley** [comment]
I think we're trying to at least get an ETA communicated within the next week or so. It's OK to leave as is for now.

**Comment by mcous** [comment]
Awesome, thanks for the help. I'll open a new PR with your listed changes after this one is merged and documentation is locked down / released.

File: devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy [hunk 7]
```diff
@@ -1,12 +1,9 @@
...
- *  Author: Keen Home
- *  Date: 2015-06-23
- */
+// keen home smart vent
+// http://www.keenhome.io
+// SmartThings Device Handler v1.0.0
 
 metadata {
-    definition (name: "Keen Home Smart Vent", namespace: "Keen Home", author: "Gregg Altschul") {
+    definition (name: "Keen Home Smart Vent", namespace: "keenhome", author: "Keen Home") {
```
Comment (dkirker) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy]: @workingmonk  Do you know how this will affect everything? It looks like the path and namespace won't match. I suspect regardless only new installs will get the new code, and existing devices will not.

Comment (mcous) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy]: Oh oops I didn't mean to change the namespace. I was just going off our GitHub org. Let me know if you need me to revert the change.

Comment (workingmonk) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy]: because it under the directory keen-home, you cannot change the namespace to "keenhome"

edit: i meant NOT. you will have to update that change

Comment (dkirker) [inline devicetypes/keen-home/keen-home-smart-vent.src/keen-home-smart-vent.groovy]: @mcous Yes, please do one more commit setting it back. Then we should be good!

**Comment by mcous** [comment]
@dkirker @workingmonk Inadvertent namespace change reverted
